coins/coins.go

```diff
@@ -0,0 +1,252 @@
+	return a.Decimal().Text('f')
+}
+
+// AsMonero converts the piconero PiconeroAmount into standard units
+func (a *PiconeroAmount) AsMonero() *apd.Decimal {
+	xmrAmt := new(apd.Decimal).Set(a.Decimal())
+	decreaseExponent(xmrAmt, NumMoneroDecimals)
+	_, _ = xmrAmt.Reduce(xmrAmt)
+	return xmrAmt
+}
+
+// WeiAmount represents some amount of ether in the smallest denomination (wei)
+type WeiAmount apd.Decimal
+
+// NewWeiAmount converts the passed *big.Int representation of a
+// wei amount to the WeiAmount type. The returned value is a copy
+// with no references to the passed value.
+func NewWeiAmount(amount *big.Int) *WeiAmount {
+	a := new(apd.BigInt).SetMathBigInt(amount)
+	return ToWeiAmount(apd.NewWithBigInt(a, 0))
+}
+
+// Decimal exists to reduce ugly casts
+func (a *WeiAmount) Decimal() *apd.Decimal {
+	return (*apd.Decimal)(a)
+}
+
+// UnmarshalText hands off JSON decoding to apd.Decimal
+func (a *WeiAmount) UnmarshalText(b []byte) error {
+	err := a.Decimal().UnmarshalText(b)
+	if err != nil {
+		return err
+	}
+	if a.Negative {
+		return errNegativeWei
+	}
+	return nil
+}
+
+// MarshalText hands off JSON encoding to apd.Decimal
+func (a *WeiAmount) MarshalText() ([]byte, error) {
+	return a.Decimal().MarshalText()
+}
+
+// ToWeiAmount casts an *apd.Decimal to *WeiAmount
+func ToWeiAmount(wei *apd.Decimal) *WeiAmount {
+	return (*WeiAmount)(wei)
+}
+
+// EtherToWei converts some amount of standard ether to an WeiAmount.
+func EtherToWei(ethAmt *apd.Decimal) *WeiAmount {
+	weiAmt := new(apd.Decimal).Set(ethAmt)
+	increaseExponent(weiAmt, NumEtherDecimals)
+	if err := roundToDecimalPlace(weiAmt, weiAmt, 0); err != nil {
+		panic(err) // shouldn't be possible
+	}
+	return ToWeiAmount(weiAmt)
+}
+
+// BigInt returns the given WeiAmount as a *big.Int
+func (a *WeiAmount) BigInt() *big.Int {
+	// Passing Quantize zero for the exponent set the coefficient to the whole-number
+	// wei value. Round-half-up is used by default.
+	wholeWeiVal := new(apd.Decimal)
+	cond, err := decimalCtx.Quantize(wholeWeiVal, a.Decimal(), 0)
+	if err != nil {
+		panic(err)
+	}
+	if cond.Rounded() {
+		// We round when converting from Ether to Wei, so we shouldn't see this
+		log.Warn("Converting WeiAmount=%s to big.Int required rounding", a)
+	}
+	return new(big.Int).SetBytes(wholeWeiVal.Coeff.Bytes())
+}
+
+// AsEther returns the wei amount as ether
+func (a *WeiAmount) AsEther() *apd.Decimal {
+	ether := new(apd.Decimal).Set(a.Decimal())
+	decreaseExponent(ether, NumEtherDecimals)
+	_, _ = ether.Reduce(ether)
+	return ether
+}
+
+// AsStandard is an alias for AsEther, returning the wei amount as ether
+func (a *WeiAmount) AsStandard() *apd.Decimal {
+	return a.AsEther()
+}
+
+// String returns the wei amount as a base10 string
+func (a *WeiAmount) String() string {
+	return a.Decimal().Text('f')
+}
+
+// ERC20TokenAmount represents some amount of an ERC20 token in the smallest denomination
+type ERC20TokenAmount struct {
+	amount      *apd.Decimal
```

Comment:
same here

Reply:
`apd.Decimal` is basically an improved `big.Int` combined with an exponent. It has much better performance characteristics than `big.Int`, because it will use built-in types if the value of the coefficient is small enough. Pushing the zeros out of the coefficient and into the exponent also improves its performance. That said, we don't care much about performance. It is the fact that the math operations are better, are lossless in base10, provide any type of rounding you may want, and provide very precise error reasons that make the package nice to work with. It is maintained by Cockroach DB, which has a $5 billion valuation and is by far the best relational database I've ever used/administrated, so support-wise it seem like a safe choice. If we are going to use `apd.Decimal` for ETH and developers have to learn the basics of it anyways, we are best off using it it for WEI, XMR and all the other currencies was my opinion. If we stored WEI as a `big.Int`, we'd still have to convert it to `apd.Decimal` before doing any kind of calculation with the exchange rate or even converting it back into ETH. Play around with it some. After learning the basics, you'll probably agree. If you have question, I can probably ask answer them quickly.